tfe/resource_tfe_registry_module.go

```diff
@@ -59,39 +67,109 @@ func resourceTFERegistryModule() *schema.Resource {
 			},
+			"namespace": {
+				Type:         schema.TypeString,
+				Computed:     true,
+				Optional:     true,
+				ForceNew:     true,
+				RequiredWith: []string{"registry_name"},
+			},
+			"registry_name": {
+				Type:         schema.TypeString,
+				Computed:     true,
+				Optional:     true,
+				ForceNew:     true,
+				RequiredWith: []string{"module_provider"},
+				ValidateFunc: validation.StringInSlice(
+					[]string{"private", "public"},
+					true,
+				),
+			},
 		},
 	}
 }
 
-func resourceTFERegistryModuleCreate(d *schema.ResourceData, meta interface{}) error {
+func resourceTFERegistryModuleCreateWithVCS(v interface{}, meta interface{}) (*tfe.RegistryModule, error) {
 	tfeClient := meta.(*tfe.Client)
-
-	// Create a new options struct.
+	// Create module with VCS repo configuration block.
 	options := tfe.RegistryModuleCreateWithVCSConnectionOptions{}
+	vcsRepo := v.([]interface{})[0].(map[string]interface{})
 
-	// Get and assert the VCS repo configuration block.
-	if v, ok := d.GetOk("vcs_repo"); ok {
-		vcsRepo := v.([]interface{})[0].(map[string]interface{})
-
-		options.VCSRepo = &tfe.RegistryModuleVCSRepoOptions{
-			Identifier:        tfe.String(vcsRepo["identifier"].(string)),
-			OAuthTokenID:      tfe.String(vcsRepo["oauth_token_id"].(string)),
-			DisplayIdentifier: tfe.String(vcsRepo["display_identifier"].(string)),
-		}
+	options.VCSRepo = &tfe.RegistryModuleVCSRepoOptions{
+		Identifier:        tfe.String(vcsRepo["identifier"].(string)),
+		OAuthTokenID:      tfe.String(vcsRepo["oauth_token_id"].(string)),
+		DisplayIdentifier: tfe.String(vcsRepo["display_identifier"].(string)),
 	}
 
 	log.Printf("[DEBUG] Create registry module from repository %s", *options.VCSRepo.Identifier)
 	registryModule, err := tfeClient.RegistryModules.CreateWithVCSConnection(ctx, options)
 	if err != nil {
-		return fmt.Errorf(
+		return nil, fmt.Errorf(
 			"Error creating registry module from repository %s: %w", *options.VCSRepo.Identifier, err)
 	}
+	return registryModule, nil
+}
+
+func resourceTFERegistryModuleCreateWithoutVCS(meta interface{}, d *schema.ResourceData) (*tfe.RegistryModule, error) {
+	tfeClient := meta.(*tfe.Client)
+	// Create module without VCS.
+	options := tfe.RegistryModuleCreateOptions{
+		Name:     tfe.String(d.Get("name").(string)),
+		Provider: tfe.String(d.Get("module_provider").(string)),
+	}
+
+	if registryName, ok := d.GetOk("registry_name"); ok {
+		options.RegistryName = tfe.RegistryName(registryName.(string))
+
+		namespace, ok := d.GetOk("namespace")
+		if registryName.(string) == "private" && ok {
+			// Namespace is not allowed for private registry name
+			return nil, fmt.Errorf("Namespace is not allowed for %s registry name", registryName)
+		}
+
+		if registryName.(string) == "public" {
+			if !ok { // Namespace is required for public registry name
+				return nil, fmt.Errorf("Namespace is required for %s registry name", registryName)
+			}
+			options.Namespace = namespace.(string)
+		}
+	}
```

Comment:
Are these conditions already handled by RegistryModules.Create ? It might simplify things to just delegate this kind of validation to the client.